landlab/components/overland_flow/generate_overland_flow_implicit_kinwave.py

```diff
@@ -0,0 +1,320 @@
+    works as follows. At each time step, we iterate from upstream to downstream
+    over the topography. Because we are working downstream, we can assume that
+    we know the total water inflow to a given cell. We solve the following mass 
+    conservation equation at each cell:
+        
+        $(H^{t+1} - H^t)/\Delta t = Q_{in}/A - Q_{out}/A + R$
+        
+    where $H$ is water depth, $t$ indicates time step number, $\Delta t$ is
+    time step duration, $Q_{in}$ is total inflow discharge, $Q_{out}$ is total
+    outflow discharge, $A$ is cell area, and $R$ is local runoff rate 
+    (precipitation minus infiltration; could be negative if runon infiltration
+    is occurring).
+    
+    The specific outflow discharge leaving a cell along one of its faces is:
+        
+        $q = (1/C_r) H^\alpha S^{1/2}$
+    
+    where $C_r$ is a roughness coefficient (such as Manning's n), $\alpha$ is
+    an exponent equal to 5/3 for the Manning equation and 3/2 for the Chezy
+    family, and $S$ is the downhill-positive gradient of the link that crosses
+    this particular face. Outflow discharge is zero for links that are flat or
+    "uphill" from the given node. Total discharge out of a cell is then the
+    sum of (specific discharge x face width) over all outflow faces
+        
+        $Q_{out} = \sum_{i=1}^N (1/C_r) H^\alpha S_i^{1/2} W_i$
+        
+    where $N$ is the number of outflow faces (i.e., faces where the ground
+    slopes downhill away from the cell's node), and $W_i$ is the width of face
+    $i$.
+    
+    We use the depth at the cell's node, so this simplifies to:
+
+        $Q_{out} = (1/C_r) H'^\alpha \sum_{i=1}^N S_i^{1/2} W_i$
+        
+    We define $H$ in the above as a weighted sum of the "old" (time step $t$)
+    and "new" (time step $t+1$) depth values:
+        
+        $H' = w H^{t+1} + (1-w) H^t$
+    
+    If $w=1$, the method is fully implicit. If $w=0$, it is a simple forward
+    explicit method.
+    
+    When we combine these equations, we have an equation that includes the
+    unknown $H^{t+1}$ and a bunch of terms that are known. If $w\ne 0$, it is
+    a nonlinear equation in $H^{t+1}$, and must be solved iteratively. We do
+    this using a root-finding method in the scipy.optimize library.
+
+    Construction:
+
+        KinwaveImplicitOverlandFlow(grid, precip_rate=1.0,
+                                    precip_duration=1.0,
+                                    infilt_rate=0.0,
+                                    roughness=0.01, **kwds)
+
+    Parameters
+    ----------
+    grid : ModelGrid
+        A Landlab grid object.
+    precip_rate : float, optional (defaults to 1 mm/hr)
+        Precipitation rate, mm/hr
+    precip_duration : float, optional (defaults to 1 hour)
+        Duration of precipitation, hours
+    infilt_rate : float, optional (defaults to 0)
+        Maximum rate of infiltration, mm/hr
+    roughnes : float, defaults to 0.01
```

Comment:
Do we want to support spatially values for quantities like precipitation rate, infiltration rate, and roughness? 

I think this would just require checking that they are either single valued quantities or at node quantities. 

Other than that this looks good to me?